First attempt at solving #50. (#62)
ietf-wg-privacypass/base-drafts: draft-ietf-privacypass-protocol.md

  store[server.id].push(tokens)
~~~

Note that the first round of the protocol is only necessitated for

> certain ciphersuites that require prior state generation. When such

Reviewer: ```suggestion
certain ciphersuites that require client and servers commit to some state. When such
```

state `st` is generated and sent to the client, the client returns the

> `st` with the `IssuanceRequest` message. The server must check that the

Reviewer: ```suggestion
`st` with the `IssuanceRequest` message. The server MUST check that the
```

> state corresponds to the `st` that was previously sent state. This

Reviewer: ```suggestion
state corresponds to `st` that was previously committed. This
```

> requires architecture for remaining stateful, or by passing the state as
> an encrypted (and authenticated) blob. Such statefulness must be
> implemented in similar ways to how TLS session resumption is managed
> {{RFC8446}}.

Reviewer: ```suggestion
requires the commitment to either be a reference to some state on the server, 
or the commitment be an encrypted (and authenticated) blob that the server can
use to recover state. The mechanism by which servers handle this state is implementation
specific, and similar to how TLS session resumption state is managed; see {{RFC8446}}
for details.
```

When the server does not need to generate this state, the client instead
DOES NOT send the CommitRequest message, and runs:

~~~

> cInput = Generate(m, null)

Reviewer: ```suggestion
cInput = Generate(m, "")
```

~~~

> A server that is expecting some non-null `st` to be passed must abort

Reviewer: ```suggestion
A server that is expecting some non-empty `st` to be passed must abort
```

> the protocol on receiving a request containing a null `st` value.

Reviewer: ```suggestion
the protocol on receiving a request containing an empty `st` value.
```

Note: currently, no ciphersuites are supported that support working with

> non-null state messages.

Reviewer: ```suggestion
empty state messages.
```

  store[server.id].push(tokens)
~~~

Note that the first round of the protocol is only necessitated for
certain ciphersuites that require client and servers commit to some
value. When such commitment `com` is generated and sent to the client,
the client returns `com` with the `IssuanceRequest` message. The server

> MUST check that the commitment corresponds to `com` that was previously

Reviewer 2: When the commitment does not corresponds to `com` that was previously
committed, should this define a specific error that we need to add to the `Error types`?

Author: @claucece Good point! Yes it should be.

opaque PrivateKey<1..2^16-1>
~~~

### CommitRequest {#pp-cli-commit-request}

The `CommitRequest` struct is simply a fixed message corresponding to
the ID of the server.

~~~
struct {

>   opaque server_id<1..2^16-1>

Reviewer: Interesting -- do we need this? Can it be empty?

Author: I was allowing for the possibility that the client may only want to allow commitments for certain types (e.g. it may only want to allow server S with ciphersuite `c1`, rather than `c2`). The `server_id` was kind of a placeholder for that. However, it is pretty ugly right now, so I would be happy to remove it.

Reviewer: 👍 Eh, it's probably fine for now.

} CommitRequest;
~~~

### CommitResponse {#pp-cli-commit-response}

The `CommitResponse` struct is contains an opaque set of bytes that

> correspond to some state that the server has generated. How this
> sequence of bytes is generated corresponds to the ciphersuite and
> whether the server is stateful or not.

Reviewer: ```suggestion
correspond to some state that the server has generated. The structure
and format of this value is implementation specific depending on whether
the server is stateful.
```

in the group. Note that the actual security level MAY be lower. See the
security considerations in {{I-D.irtf-cfrg-voprf}} for examples.

The COMMIT parameter refers to whether the first round of the issuance

> phase of the protocol is necessary. When this set to false, the client

Reviewer: ```suggestion
phase of the protocol is necessary. When this is false, the client
```

> ignores the first message and uses a null value for the state parameter
> sent by the server.

Reviewer: ```suggestion
ignores the first message and uses an empty value for the state parameter
to `Generate`.
```
